hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataResolver.java

```diff
@@ -56,59 +58,111 @@ public Mapping resolve(String iMapName) {
-        Metadata metadata = hd ? resolveFromHd(container) : resolveFromHeap(iMapName, context);
+        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE
+                && !container.getMapConfig().getTieredStoreConfig().isEnabled();
+
+        Metadata metadata = hd
+                ? resolveFromContentsHd(iMapName, context)
+                : resolveFromContents(iMapName, context);
         return metadata == null
                 ? null
                 : new Mapping(iMapName, iMapName, null, TYPE_NAME, null, metadata.fields(), metadata.options());
     }
 
     @Nullable
-    @SuppressWarnings("rawtypes")
-    private Metadata resolveFromHd(MapContainer container) {
-        if (container.getIndexes() == null) {
-            return null;
+    private Metadata resolveFromContents(String name, MapServiceContext context) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
+            Metadata resolved = resolveFromContentsRecordStore(nodeEngine, recordStore);
+            if (resolved != null) {
+                return resolved;
+            }
         }
+        return null;
+    }
 
-        InternalIndex[] indexes = container.getIndexes().getIndexes();
-        if (indexes == null || indexes.length == 0) {
-            return null;
+    private Metadata resolveFromContentsHd(String name, MapServiceContext context) {
+        // Iterate only over local partitions.
+        // MC will invoke this operation on each member until it finds some data.
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            // HD access must be from partition threads.
+            GetAnyMetadataOperation op = new GetAnyMetadataOperation(name);
+            Metadata resolved = invoke(op, partitionContainer.getPartitionId());
+            if (resolved != null) {
+                return resolved;
+            }
         }
+        return null;
+    }
 
-        InternalIndex index = indexes[0];
-        Iterator<QueryableEntry> entryIterator = index.getSqlRecordIterator(false);
-        if (!entryIterator.hasNext()) {
-            return null;
+    private static final class GetAnyMetadataOperation extends Operation {
+        private final String mapName;
+
+        private GetAnyMetadataOperation(String mapName) {
+            this.mapName = mapName;
         }
 
-        QueryableEntry entry = entryIterator.next();
-        return resolveMetadata(entry.getKey(), entry.getValue());
+        @Override
+        public void run() {
+            MapService service = getNodeEngine().getService(MapService.SERVICE_NAME);
+            MapServiceContext context = service.getMapServiceContext();
+            RecordStore<?> recordStore = context.getExistingRecordStore(getPartitionId(), mapName);
```

Comment:
Not introduced this PR but I'm seeing so many IMap internals leaked in MetadataResolver which is in different module already. Maybe we can move this metadata resolution logic inside a new MutationObserver, we do something similar for json metadata. 
And then, sql module can reach that metadata to generate what needed for it. By this way, we don't need to have different mechanism for heap or HD data.

Reply:
`MetadataResolver` needs only 1 IMap entry in serialized form. It must be serialized, not deserialized, because from `Data` we can get serialization format which is needed for `CREATE MAPPING`. Is there a simpler way to do it?

`MutationObserver` is invoked only for updates, but `MetadataResolver` is invoked on MC user request, when there are no updates. And it does not make sense to observe all IMaps for all the time.

`JsonMetadata` seems like special type of record, we need to operate on any kind of IMap.

Reply:
>MetadataResolver needs only 1 IMap entry in serialized form. It must be serialized, not deserialized, because from Data we can get serialization format which is needed for CREATE MAPPING. Is there a simpler way to do it?

With a `MutationObserver` you can do this, it can have state for only 1 entry, and MC can read it. 
Raw data  is passed without deserialization to `MutationObserver`.